### ISSUE TITLE
vmstorage failing to answer vmselect/vminsert requests for an extended period after restart  (LTS, v1.87.7-cluster)

### ISSUE THREAD

**Issue by nick-kentik** [VictoriaMetrics/VictoriaMetrics#5043]
### Describe the bug

We've implemented rolling restarts for our VictoriaMetrics cluster (7 vmstorage, 42 vminsert, 42 vmselect, spread across 7 nodes) in an attempt to get zero-downtime upgrades or configuration changes.

When restarting `vmstorage` instances, we do so one at a time, waiting for each `vmstorage` to come back up, checking its health via the `/health` endpoint (also giving it at least 30 seconds to stabilise before moving to the next one).

We have `-replicationFactor 2` configured on vminsert and vmselect; each reports up to 4 vmstorage nodes being unavailable during a rolling restart, however, so we see a period of queries and inserts failing, leading to various alerts being raised (we use Grafana alerting) and data eventually being lost (as Telegraf write buffers fill up).

Looking into the logs of a `vmstorage` during its restart, there is a 20-minute period immediately following the restart where `vminsert` and `vmselect` connections are accepted, but not responded to - errors vary from `EPIPE`, `EOF` or `i/o timeout` through queries failing with `context: deadline exceeded`.

Interspersed into these failures are 81 different "merged (..." messages. I've broken them up according to the type, and see:

- 2 merging partition (just a filename - e.g. `"4347187_3850839_20230921093651.238_20230921094315.000"`) - are these `big` merges?
- 14 merging partition `"/storage/data/small/..."`
- 17 merging table `"/storage/indexdb/..."`
- 17 merging table `""`

I feel like what's happening is similar to https://github.com/VictoriaMetrics/VictoriaMetrics/issues/2211 , but with a vmstorage restart *causing* (rather than solving) the issue. During this 20-minute period, the `vmstorage` is writing hard to the disk, e.g.:

![Screenshot from 2023-09-21 12-25-14](https://github.com/VictoriaMetrics/VictoriaMetrics/assets/107049349/c0af0c21-e253-42e5-b352-c43e98100071)

These are 64-core hosts and system load goes up to ~60. There's plenty of free RAM and the CPU isn't loaded, so I assume we're bound on write I/O. What I don't understand is why `vmstorage` decides to do all this work immediately on a restart, instead of spreading it out.

I know about the `-bigMergeConcurrency` option, but I haven't tried it out yet as I think most of these operations aren't big merges, and so the flag wouldn't have an effect. It's also not clear to me what setting would be good, given the hardware. I'd much rather trust the defaults.

We see 25 "assisted merges" immediately following the restart. We don't see any assisted merges at any other time, so I don't think this is a case of resource exhaustion as suggested in https://github.com/VictoriaMetrics/VictoriaMetrics/issues/2836

The comments about assisted merges blocking queries are interesting though (although we see ingest blocked as well). 

### To Reproduce

I have the start of a theory. On the graceful shutdown of `vmstorage`, immediately before it is started again, lots of data gets flushed to disk. Perhaps this is lots of small parts?

On large clusters, perhaps this operation triggers the "assisted merge required" logic, leading to queries being blocked until those parts are all merged together?

The high disk I/O that creates would perhaps be enough to explain why ingest begins to fail?

### Version

```
/ # /vmstorage-prod -version
vmstorage-20230812-211856-tags-v1.87.7-cluster-0-ga783d88886
```

## Distribution of merge messages following a `vmstorage` restart

![Screenshot from 2023-09-21 17-51-21](https://github.com/VictoriaMetrics/VictoriaMetrics/assets/107049349/c6119769-05ff-4e32-816e-fc76a1ef6257)

## VictoriaMetrics cluster dashboard
![Screenshot from 2023-09-21 18-02-25](https://github.com/VictoriaMetrics/VictoriaMetrics/assets/107049349/9dd2f14f-2cfe-4c80-840f-bbd310275904)
![Screenshot from 2023-09-21 18-02-11](https://github.com/VictoriaMetrics/VictoriaMetrics/assets/107049349/f359b029-984a-4242-8c57-a34ecd2de59b)
![Screenshot from 2023-09-21 18-02-02](https://github.com/VictoriaMetrics/VictoriaMetrics/assets/107049349/4f4849fe-8628-4294-8e32-17d1bb92753f)
![Screenshot from 2023-09-21 18-01-56](https://github.com/VictoriaMetrics/VictoriaMetrics/assets/107049349/9ccdd638-ae3a-4eb7-84ef-55f15835d76e)
![Screenshot from 2023-09-21 18-01-40](https://github.com/VictoriaMetrics/VictoriaMetrics/assets/107049349/66621c21-5317-43f5-84f8-f02800cee480)
![Screenshot from 2023-09-21 18-01-29](https://github.com/VictoriaMetrics/VictoriaMetrics/assets/107049349/59b36af2-83a5-4bb1-9d40-3e3a8267e6b2)
![Screenshot from 2023-09-21 18-01-16](https://github.com/VictoriaMetrics/VictoriaMetrics/assets/107049349/f76cd086-fe22-4655-afd3-b68d037fdc6f)
![Screenshot from 2023-09-21 18-01-07](https://github.com/VictoriaMetrics/VictoriaMetrics/assets/107049349/574e9d4f-8f64-41f9-bd72-e781f015f6dc)

![Screenshot from 2023-09-21 18-06-28](https://github.com/VictoriaMetrics/VictoriaMetrics/assets/107049349/ef6e4686-348c-470f-9d79-b6aabc64ea79)
![Screenshot from 2023-09-21 18-06-06](https://github.com/VictoriaMetrics/VictoriaMetrics/assets/107049349/a6f5d324-36b7-4653-a679-092f25568249)

**Comment by nick-kentik** [comment]
(I'd love to upgrade to v1.87.9 but nothing in the changelog looks immediately relevant, and it's difficult to roll out upgrades while the zero-downtime upgrade isn't working correctly :sweat_smile: . I could go for a manual one-by-one upgrade if you think there's a chance something in .8 or .9 would fix it, though)

**Comment by nick-kentik** [comment]
Hi @hagen1778, thanks for the reply. I read through #4922 and it does seem to have a lot to say about improving the vminsert situation. Is any of this planned for backup to 1.87? Or the indexdb organisation work you linked to? We're not absolutely set on staying on LTS, but it's our goal at the moment.

We have node-exporter, and psi metrics too, and they both show a large spike in CPU and disk I/O (but not I/O wait). What I'm seeing is that vmstorage maxed out CPU merging the various parts - but it did so voluntarily, and unnecessarily (AFAICT, since vmstorage was working fine with the parts on disk before restart), on startup, for 20 straight minutes - which is really the worst possible time for it to decide to do that.

If it deferred all this work to a random time in the future, for instance, it wouldn't matter as much since it would only become service-affecting if two vmstorage instances happened to choose the same timeslot to do it in.

I investigated the "shutdown writes lots of small parts to disk" theory a bit more and didn't find strong evidence for it - after an uptime of a couple of days, a `tree` diff didn't show lots of parts being written to disk on shutdown, and on startup there weren't a large number of merges triggered. Perhaps it's more an accumulation of parts over time? Maybe vmstorage needs to merge more aggressively while running in our setup?

I'm going to do a bump to v1.87.9 today, and will monitor each vmstorage before moving on to the next,so I'll be able to report on how many (if any) exhibit it after around a week of uptime. I'll take `tree` diffs as well; hoping to catch one in the act.

5 minutes between vmstorage instances is about doable for a single cluster, but we're running ~10 separate ones and like to upgrade them in series, rather than in parallel, so it quickly becomes infeasible. We're only seeing this on the largest cluster at the moment though. Really, we want to trust `vmstorage`'s health endpoint.

**Comment by nick-kentik** [comment]
OK, @hagen1778 , I've gone through and done a rolling upgrade of everything to v1.87.9. It was smooth and we didn't see the merge operation storm that caused an outage a week ago.

We've had two outages like it - I opened this ticket on the second one - separated by ~2 weeks (30th august and 21st September), whereas this upgrade was only a week after the last restart, so it might be something that only happens for us on a restart after extended uptim / accumulation of trash.

We didn't see any insert issues during this most recent upgrade, so I think it's pretty clear we have the capacity to handle load with just 6 vmstorages up, rather than the usual 7.

Zooming out on vmstorage's merge stats as far as we have data on them, I see this:

![Screenshot from 2023-09-28 15-59-40](https://github.com/VictoriaMetrics/VictoriaMetrics/assets/107049349/33dc111d-f264-4d7a-bbc8-865379d0f3ff)

Which shows just how abnormal the two tallest spikes (correlating to vmstorage restarts leading to outages) are. It's a huge number of `indexdb/inmemory` merges happening immediately after a restart, for... some reason. I don't know at the moment what those operations correspond to; I'll hopefully get to do some digging on it tomorrow.

Zooming in on `indexdb/inmemory`, split out by host, over those two events:

![Screenshot from 2023-09-28 16-04-38](https://github.com/VictoriaMetrics/VictoriaMetrics/assets/107049349/55728b3b-dd6f-4654-9e00-e5330016c66e)

![Screenshot from 2023-09-28 16-05-26](https://github.com/VictoriaMetrics/VictoriaMetrics/assets/107049349/df2f4d11-e270-4e7e-925e-bf1ba908fcfa)

One thing we saw in our metrics was that memory psi went up significantly - e.g. vmstorage was waiting for memory allocations to complete in the kernel:

![Screenshot from 2023-09-28 16-08-54](https://github.com/VictoriaMetrics/VictoriaMetrics/assets/107049349/e18b970c-fe24-48a7-8971-90f083c9c0a2)

I don't have any conclusions yet, but it's nice to be able to point confidently at a specific thing causing the load. It does look like spacing out the restarts over a longer period of time would avoid an outage even when we manage to trigger this high-load operation on startup; but it does feel wrong that it's happening at-all on startup, and that vmstorage advertises itself as healthy when it isn't.

**Comment by wjordan** [comment]
I saw very similar behavior as described in this issue, and in my case I managed to narrow the issue down to two different root causes:

1. Rerouting storm, as described in #4922. I've been experimenting with delaying rerouting by leveraging large insert buffers as described in https://github.com/VictoriaMetrics/VictoriaMetrics/issues/4922#issuecomment-1762070962, and my initial results are showing some significant improvements to recovery and stabilization times across vmstorage service restarts and intermittent network issues.
2. Improperly written cache files after unclean shutdown. My cluster initially had a 5-second shutdown timeout, which turned out to be too short for some vmstorage nodes to properly finish writing out the `curr_hour_metric_ids` file to the filesystem which it only does during a graceful shutdown. As a result, vmstorage invalidated the entire metric-id cache and had to rebuild this cache entirely from scratch on every restart, which took several minutes to complete. Errors such as these appeared in the log, indicating the issue:
    ```
     discarding /vmstorage-data/cache/curr_hour_metric_ids, since it contains outdated hour; got 471015; want 471550
    ```
    Also, the `curr_hour_metric_ids` file itself was only 24 bytes and had a modified date several months old, indicating a problem. (Properly behaving storage nodes had ~100MB files.)
    After adjusting the shutdown timeout to 30 seconds, restarts were quick and the error no longer appeared in my log.
